### PR TITLE
add sample_logits  and sampled_softmax_with_cross_entropy op

### PR DESCRIPTION
**Pull request by xuezhong**
add sample_logits op

### CODE REVIEW

**Comment by kuke** [comment]
Please add `test=develop` in commit message

File: paddle/fluid/operators/sample_logits_op.cc [hunk 1]
```diff
@@ -0,0 +1,225 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc:1]: 2016->2019

File: paddle/fluid/operators/sample_logits_op.cc [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+             "(Tensor, default: Tensor<float>), The unscaled log probabilities "
+             "which is a 2-D tensor with shape [N x K]. N is the batch_size, "
+             "and K is the class number.");
+    AddInput("Label",
+             "(Tensor) The ground truth which is a 2-D tensor. Label is a "
+             "Tensor<int64> with shape [N x NT], where NT is the number of"
+             "true labels for each example.");
+    AddInput(
+        "CustomSamples",
+        "(Tensor, default: Tensor<int64_t>), A 2-D tensor with shaoe [N x "
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: shaoe->shape

File: paddle/fluid/operators/sample_logits_op.cc [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        "S+NT]."
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: 如果` with true labels at first`的话，这里写成`NT + S`更合适吧？`S`可以进一步解释下么？

File: paddle/fluid/operators/sample_logits_op.cc [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        "The customized sample labels with true labels at first. This tensor"
+        "is only use_custom_samples is true.")
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: 这句话好像不通顺

File: paddle/fluid/operators/sample_logits_op.cc [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        .AsDispensable();
+    AddInput(
+        "CustomProbabilities",
+        "(Tensor, default: Tensor<float>), A 2-D tensor with shaoe [N x S+NT]."
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: shaoe-> shape,
[N x S+NT]->[N, S+NT]

File: paddle/fluid/operators/sample_logits_op.cc [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        "The customized sample probabilities with true labels at first. This "
+        "tensor is only use_custom_samples is true.")
+        .AsDispensable();
+    AddOutput(
+        "Samples",
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: 这里是sampled labels的意思吧？

File: paddle/fluid/operators/sample_logits_op.cc [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        "(Tensor, default: Tensor<int64_t>), A 2-D tensor with shape [N x "
+        "S+NT]."
+        "The outputs value of sampler by given the true label, where S is the "
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: `by given` 好像不能这么用吧？

File: paddle/fluid/operators/sample_logits_op.cc [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        "labels and S negative labels for each example. This will be used in"
+        "backward calculation.")
+        .AsIntermediate();
+    AddOutput(
+        "Probabilities",
+        "(Tensor, default: Tensor<float>), A 2-D tensor with shape [N x "
+        "S+NT]."
+        "The outputs value of progabilites of samples by given the true label, "
+        "where S is the "
+        "number of negative sample for each example. So Samples includes NT "
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: >So Samples includes NT

这里已经不是Samples了

File: paddle/fluid/operators/sample_logits_op.cc [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+        "true"
+        "labels and S negative labels for each example. This will be used in"
+        "backward calculation.")
+        .AsIntermediate();
+    AddOutput(
+        "Probabilities",
+        "(Tensor, default: Tensor<float>), A 2-D tensor with shape [N x "
+        "S+NT]."
+        "The outputs value of progabilites of samples by given the true label, "
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: 这里的samples是sampled labels的意思么？
progabilites -> probabilities
The outputs value of probabilities 改成 The probabilities ？

File: paddle/fluid/operators/sample_logits_op.cc [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        "where S is the "
+        "number of negative sample for each example. So Samples includes NT "
+        "true"
+        "labels and S negative labels for each example.")
+        .AsIntermediate();
+    AddOutput("SampledLogits",
+              "(Tensor, default: Tensor<float>), A 2-D tensor with shape"
+              "[N x S+NT]. The outputs value of sample logits, which will be"
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: [N x S+NT]->[N , S+NT]

sample logits -> sampled logits ?

File: paddle/fluid/operators/sample_logits_op.cc [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+              "used in backward calculation.")
+        .AsIntermediate();
+    AddOutput(
+        "SampledLabel",
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: 这个和Output(Samples)不一样？

File: paddle/fluid/operators/sample_logits_op.cc [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+        "(Tensor, default: Tensor<int64>), A 2-D tensor. The sampled label"
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: label->labels

File: paddle/fluid/operators/sample_logits_op.cc [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+        "with shape [N x S + NT].");
+    AddAttr<bool>(
+        "use_custom_samples",
+        "An indicator whether to use custom samples with probabilities, if True"
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: custom-> customized
其它地方也要改下

File: paddle/fluid/operators/sample_logits_op.cc [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+        "the operator will use custom samples and custom probabilities"
+        "otherwise, the operator will generate them by itself.")
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc:79]: otherwise前边少一个标点符号吧

File: paddle/fluid/operators/sample_logits_op.cc [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+        "An indicator whether to remove accidental hits when samples hits true"
+        "labels, the removal is implemented by subtracting the corresponding"
+        "logits by float_max to subpress their softmax to be zero.")
+        .SetDefault(true);
+    AddAttr<int>("num_samples", "The number of negative samples.");
+    AddAttr<int>("seed", "Random seed for generating samples").SetDefault(0);
+
+    AddComment(R"DOC(
+  """
+  Computes sampled output training logits and labels suitable for implementing
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc:98]: 最好加个例子解释下

File: paddle/fluid/operators/sample_logits_op.cc [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+    ctx->SetOutputDim("SampledLabel", {logits_dims[0], labels_dims[1]});
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
+      const framework::ExecutionContext& ctx) const override {
+    auto data_type = framework::GetDataTypeOfVar(ctx.InputVar("Logits"));
+    framework::OpKernelType kt =
+        framework::OpKernelType(data_type, ctx.device_context());
+    // kt.place_ = platform::CPUPlace();
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.cc]: 去掉注释语句

File: paddle/fluid/operators/sample_logits_op.h [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+    // device contexts
+    auto& dev_ctx =
+        context.template device_context<platform::CPUDeviceContext>();
+
+    // UNDERSTAND: allocate memories for temporaries
+    sampled_logits->mutable_data<T>(samples_dim, context.GetPlace());
+    auto sampled_label_data =
+        sampled_label->mutable_data<int64_t>(label_dim, context.GetPlace());
+    for (int i = 0; i < batch_size; ++i)
+      for (int j = 0; j < num_true; ++j)
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/sample_logits_op.h]: for循环最好加上大括号

**Review by wanghaoshuang — CHANGES_REQUESTED** [review]

File: paddle/fluid/operators/sample_logits_op.cc (hunk 3 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc]: ok 这个地方实际上是N x (NT + S)

File: paddle/fluid/operators/sample_logits_op.cc (hunk 2 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc]: ok

File: paddle/fluid/operators/sample_logits_op.cc (hunk 4 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc]: ok

File: paddle/fluid/operators/sample_logits_op.cc (hunk 5 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc]: ok

File: paddle/fluid/operators/sample_logits_op.cc (hunk 14 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc:79]: ok

File: paddle/fluid/operators/sample_logits_op.cc (hunk 15 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc:98]: 这个op不进入layers 因此文档主要在sampled_softmax上。

File: paddle/fluid/operators/sample_logits_op.cc (hunk 16 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc]: ok

File: paddle/fluid/operators/sample_logits_op.h (hunk 17 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.h]: ok

File: paddle/fluid/operators/sample_logits_op.cc (hunk 6 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc]: 包括正负样本

File: paddle/fluid/operators/sample_logits_op.cc (hunk 7 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc]: ok

File: paddle/fluid/operators/sample_logits_op.cc (hunk 8 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc]: 应该是正负样本

File: paddle/fluid/operators/sample_logits_op.cc (hunk 9 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc]: ok

File: paddle/fluid/operators/sample_logits_op.cc (hunk 12 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc]: ok

File: paddle/fluid/operators/sample_logits_op.cc (hunk 13 above)
Comment (xuezhong) [reply paddle/fluid/operators/sample_logits_op.cc]: ok

**Review by wanghaoshuang — DISMISSED** [review]
LGTM.

File: python/paddle/fluid/tests/unittests/testsuite.py [hunk 18]
```diff
@@ -156,8 +156,26 @@ def create_var(block, name, np_list, var_proto):
...
+            'in_dtype': input.dtype
+        })
+    op.desc.infer_var_type(block.desc)
+    op.desc.infer_shape(block.desc)
+    return out
+
+
 def append_loss_ops(block, output_names):
     mean_inputs = list(map(block.var, output_names))
+    mean_inputs = [var_cast(block, x) for x in mean_inputs]
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/testsuite.py]: why they must be fp32? Can you do the cast in your own op test?

File: python/paddle/fluid/tests/unittests/test_layers.py [hunk 19]
```diff
@@ -374,6 +374,17 @@ def test_im2sequence(self):
...
+    def test_sampled_softmax_with_cross_entropy(self):
+        program = Program()
+        with program_guard(program):
+            logits = layers.data(name='Logits', shape=[256], dtype='float64')
+            label = layers.data(name='Label', shape=[1], dtype='int64')
+            num_samples = 25
+            output = layers.sampled_softmax_with_cross_entropy(logits, label,
+                                                               num_samples)
+            self.assertIsNotNone(output)
+        print(str(program))
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_layers.py:386]: you are following very bad example...

**Review by panyx0718 — APPROVED** [review]

**Review by guoshengCS — APPROVED** [review]

File: python/paddle/fluid/tests/unittests/testsuite.py (hunk 18 above)
Comment (xuezhong) [reply python/paddle/fluid/tests/unittests/testsuite.py]: another pr fix this problem， remove test first
